domain/access/state/permission.go

```diff
@@ -713,3 +722,192 @@ WHERE  grant_on = $M.grant_on
+        LEFT JOIN ctrl ON p.grant_on = ctrl.c
+WHERE   u.name = $input.name
+AND     (p.grant_on = $input.grant_on OR p.grant_on = ctrl.c)
+AND     u.disabled = false
+AND     u.removed = false
+`
+	authStmt, err := st.Prepare(authQuery, dbReadUserPermission{}, input{})
+	if err != nil {
+		return apiUserUUID, errors.Trace(err)
+	}
+	var readPerm []dbReadUserPermission
+	in := input{
+		GrantOn: grantOn,
+		Name:    apiUser,
+	}
+	err = tx.Query(ctx, authStmt, in).GetAll(&readPerm)
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
+		return apiUserUUID, errors.Annotatef(accesserrors.PermissionNotValid, "on %q", grantOn)
+	} else if err != nil {
+		return apiUserUUID, errors.Annotatef(err, "verifying authorization of %q", grantOn)
+	}
+	apiUserUUID = user.UUID(readPerm[0].GrantTo)
+
+	for _, read := range readPerm {
+		if read.AccessType == string(corepermission.SuperuserAccess) ||
+			read.AccessType == string(corepermission.AdminAccess) {
+			return apiUserUUID, nil
+		}
+	}
+
+	return apiUserUUID, accesserrors.PermissionNotValid
+}
+
+func (st *PermissionState) grantPermission(ctx context.Context, tx *sqlair.TX, args access.UpsertPermissionArgs) error {
+	userAccessLevel, err := st.userAccessLevel(ctx, tx, args.Subject, args.AccessSpec.Target)
+	if err != nil {
+		return errors.Annotatef(err, "getting current access for grant")
+	}
+	aSpec := corepermission.AccessSpec{
+		Target: args.AccessSpec.Target,
+		Access: userAccessLevel,
+	}
+	grantAccess := args.AccessSpec.Access
+	if aSpec.EqualOrGreaterThan(grantAccess) {
+		return errors.Errorf("user %q already has %q access or greater", args.Subject, grantAccess)
+	}
+	if err := st.updatePermission(ctx, tx, args.Subject, args.AccessSpec.Target.Key, string(grantAccess)); err != nil {
+		return errors.Annotatef(err, "updating current access during grant")
+	}
+	return nil
+}
+
+func (st *PermissionState) revokePermission(ctx context.Context, tx *sqlair.TX, args access.UpsertPermissionArgs) error {
+	newAccess := args.AccessSpec.RevokeAccess()
+	if newAccess == corepermission.NoAccess {
+		err := st.deletePermission(ctx, tx, args.Subject, args.AccessSpec.Target)
+		return errors.Annotatef(err, "revoking %q", args.AccessSpec.Access)
+	}
+	if err := st.updatePermission(ctx, tx, args.Subject, args.AccessSpec.Target.Key, string(newAccess)); err != nil {
+		return errors.Annotatef(err, "updating current access during revoke")
+	}
+	return nil
+}
+
+func (st *PermissionState) deletePermission(ctx context.Context, tx *sqlair.TX, subject string, target corepermission.ID) error {
+	type input struct {
+		Name    string `db:"name"`
+		GrantOn string `db:"grant_on"`
+	}
+
+	// The combination of grant_to and grant_on are guaranteed to be
+	// unique, thus it is all that is deleted to select the row to be
+	// deleted.
+	deletePermission := `
+DELETE
+FROM   permission
+WHERE  uuid IN (
+       SELECT p.uuid
+       FROM   permission p
+              LEFT JOIN user AS u ON u.uuid = p.grant_to
```

Comment:
I don't think you want to `LEFT JOIN` this one. It's probably safe given the filter, but you're stating "delete all the permissions whether they have a record in the other table or not" in the sub-select.

Reply:
Actually, I think this is a more cogent statement:
```sql
DELETE FROM permission
WHERE  p.grant_on = $input.grant_on
AND    grant_to IN (
    SELECT uuid 
    FROM   user
    WHERE  $input.name
    AND    removed = false
)
```

Reply:
Why add `AND removed = false` here? Do we preserve permissions if a user has been removed? What about disabled?

Reply:
If you have the name re-used, you don't want to revoke from the wrong one. Should probably check for modified rows too.